File: isPalindromeLinkedList.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Nope


// Your code here along with comments explaining your approach
// Travers the linked list using two pointers: fast and slow and reverse the list from the mid to the end of this list
// check if the list is palindrome by traversing the given list and the reversed list
// "static void main" must be defined in a public class.
public class Main {
    
    static class ListNode{
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int v) {
            this.val = v;
            this.next = null;
        }
        
        ListNode(int v, ListNode next) {
            this.val = v;
            this.next = next;
        }
    }
    
        public static void main(String[] args) {
            
            ListNode head = new ListNode(1);
            head.next = new ListNode(2);
            head.next.next = new ListNode(3);
            head.next.next.next = new ListNode(2);
            head.next.next.next.next = new ListNode(1);
            
            System.out.println(isPalindrome(head));
        }
        
        private static boolean isPalindrome(ListNode head) {
            if(head == null || head.next == null)
                return true;
            
            ListNode slow = head;
            ListNode fast = head;
            
            while(fast!=null && fast.next!= null) {
                fast = fast.next.next;
                slow = slow.next;
            }
            
            slow = reverseList(slow);
            fast = head;
            
            while(slow != null) {
                if(fast.val != slow.val) {
                    return false;
                }
                fast = fast.next;
                slow = slow.next;
            }
            
            return true;
        }
        
        private static ListNode reverseList(ListNode head) {
            ListNode curr = head;
            ListNode prev = null;
            ListNode next = null;
            
            while(curr != null) {
                next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }
            
            return prev;
        }
    }
    

File: is-palindrome-linked-list.cpp
/*
Use two pointers (slow and fast) to find the middle of the linked list.
Reverse the second half of the linked list.
Compare the reversed second half with the first half for palindrome checking.
Time Complexity: O(N), where N is the number of nodes in the linked list. 
We traverse the list twice - once to find the middle and once to compare values.
Space Complexity: O(1), as we use a constant amount of extra space for pointers and do not use additional data structures.*/
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(!head ||!head->next)return true;
        ListNode* slow=head;
        ListNode* fast=head;
        while(fast && fast->next){
            slow=slow->next;
            fast=fast->next->next;
        }
        slow=reverseLinkedList(slow);
        while(slow){
            if(slow->val!=head->val)return false;
            head=head->next;
            slow=slow->next;
        }
        return true;
    }
private:
    ListNode* reverseLinkedList(ListNode* head){
        ListNode* prev=nullptr, *curr=head, *next=nullptr;
        while(curr){
            next=curr->next;
            curr->next=prev;
            prev=curr;
            curr=next;
        }
        return prev;
    }
};
